### DISCUSSION TITLE
Issues Installing Quantum Package Using zsh Shell

### DISCUSSION THREAD

**Discussion by ghost** [QuantumPackage/qp2#332] [Q&A] [answered]
When attempting to install Quantum Package 2 following the install instructions at https://quantum-package.readthedocs.io/en/master/intro/install.html on a 2020 MacBook Pro (default shell zsh) running Ventura, I run into several issues which seem to ultimately be related to the shell environment. I am unable to figure out how to build Quantum Package 2 due to these issues.

This is what I have tried:
**Step 1:** `git clone https://github.com/QuantumPackage/qp2 (then cd into qp2)`
**Step 2:** `./configure -i all`        (NOTE: `./configure` on its own gave me issues about dependencies not being installed)
**Step 3:** `./configure -c config/gfortran.cfg`
**Step 4:** `source quantum_package.rc`
       --> **ERROR:** `quantum_package.rc:cd:22: no such file or directory: /Users/danielstaros/etc`
**Step 5:** Set `QP_ROOT="/Users/danielstaros/Codes/qp2"` within quantum_package.rc
**Step 6:** `source quantum_package.rc`
       --> **ERROR:** `quantum_package.rc:cd:22: no such file or directory: /Users/danielstaros/etc`
**Step 7:** `echo $(ps -p $$ -ocomm=)`
       --> **OUTPUT:** `-zsh`
**Step 8:** Change `"zsh"` in quantum_package.rc to `"-zsh"`
**Step 9:** `source quantum_package.rc`
       --> **ERRORS:** `configure.rc:34: command not found: complete`
       --> `qp.rc:301: command not found: complete`
**Step 10:** Add the following to my ~/.zshrc file and re-source ~/.zshrc (suggestion from [StackOverflow](https://superuser.com/questions/1714374/command-not-found-complete-message-appears-every-time-i-open-the-terminal)):
       --> `autoload -U +X bashcompinit && bashcompinit`
       --> `autoload -U +X compinit && compinit`
**Step 11:** `source quantum_package.rc`
       --> **NO ERROR!**
**Step 12:** `make`
       --> **ERRORS:** `ocamlfind: Package 'zmq' not found`
       --> `Command exited with code 2.`
       --> `Compilation unsuccessful after building 1 target (0 cached) in 00:00:00.`
       --> `make[1]: *** [qptypes_generator.byte] Error 10`
       --> `FAILED: ocaml_target `
       --> `make -C /Users/danielstaros/Codes/qp2/ocaml`
       --> `ninja: build stopped: subcommand failed.`
       --> `make: *** [default] Error 1`

This looks like some issue regarding ocaml, but ocaml seemed to install correctly when I configured the dependencies in **Step 2**. I have not been able to troubleshoot this error, but even if I was, this example still makes it clear that installing Quantum Package 2 in a zsh shell on MacBook Pro is not particularly straightforward. I believe the user experience would definitely benefit from changes that address this issue, if possible. Any insight for how to fix this would be greatly appreciated!

Thank you,
Daniel Staros

**Comment by scemama** [comment, marked as answer]
Dear Daniel,
MacOS is not supported by QP, only Linux. We are working on it (in low priority), but there is still a lot to do.
best,
Anthony

**Comment by ghost** [comment]
Dear Anthony,

Thank you for the prompt response. I see. I am looking forward to the support whenever it is launched :) Feel free to close this issue until then.

Kind regards,
Daniel

**Comment by mveril** [comment]
Hello @djstaros,

Sorry for the big delay on my response but I not work actively on the project now. I did my PhD with the team that created Quantum Package (QP). I can confirm QP only supports Linux. However, it is possible to run Linux on macOS or Windows using lightweight VM technology.

Personally, I used QP on Windows with WSL (Windows Subsystem for Linux) and it performed well. I also developed a Quantum Package Docker container that is the core of the try quantum package on the quantum package website). Since Docker also uses lightweight VMs on Windows and Mac (with the WSL backend on Windows), it works effectively. I tested it on a Mac during my PhD, and it worked perfectly (I didn't compare the performance, though).

Another option is Multipass, which is similar to WSL on Windows (running a Linux distribution under a lightweight VM). However, since Multipass is built by Canonical, it only supports Ubuntu as the distribution choice. You can find more information about Multipass [here](https://multipass.run/).

I haven't tested Multipass, so I can't confirm if it works, but Docker works perfectly. I have prepared an updated version of QP_demo for QP 2.2.2. I created a branch which includes compilation tools for you, so you can also create Quantum Package plugins with it. You can find it at [this link](https://github.com/mveril/qp_demo/tree/2.2.2-ubu-24.04-compiled). You can also mount a local directory into the container with the `-v` argument of the `docker run` command.

So, if you really want QP on your Mac, you definitely have the Docker container option, and maybe Multipass.

**Comment by t-diata** [comment]
Hi,
I'm reopening this issue because I recently tried to install Quantum Package on a Mac and I wanted to leave a record of the steps I followed (if this helps anyone, so much the better).

**Warning**: By following these steps, you will be modifying Quantum Package source files. Any re-installation or update of the software will certainly require you to reinstall Quantum Package from scratch, repeating these steps and analyzing any other incompatibilities encountered. 

1. Check that the shell is zsh with `ps -p $$`. If not, then run `exec zsh`
2. Download the source file : `git clone https://github.com/QuantumPackage/qp2.git`
3. Install the dependencies `./configure`, then `./configure -i [the package to install]`. For example: `.\configure -i ninja`
4. **Warning**: for trexio installation. Download the file corresponding to the trexio version indicated in `./configure from github.com/TREX-CoE/trexio/releases`. Copy the file to qp2/external before running `./configure -i trexio` (or `trexio-nohdf5`)
5. Open quantum_package.rc and replace line 43 `QP_ROOT=$(dirname ${BASH_SOURCE}[0])` with `QP_ROOT="$(cd $(dirname $0) && pwd)"`
6. Add the highlighted lines in the following image to the quantum_package.rc file:
![Screenshot 2025-03-24 at 14 29 13](https://github.com/user-attachments/assets/7566def1-ca46-4e0f-9445-0cdf70c2bf7f)
_Justification_: In case the if test gives you an error, add the QP_ROOT in your PATH. zcat: Since `zcat` is not recognized in the same way on `zsh`, I tried to use an alias. But it's ignored (maybe I'm writing it in the wrong place). Solution a bit shaky (you have to do it in fact): from qp2/, I did a `grep -r zcat` and replaced all `zcat` appearances with `gunzip -c` (which is an alternative to `zcat` which works with zsh shell). (any better solution is welcomed) Files to modify: `./bin/qp_reset` and `./external/ezfio/src/ezfio.ml`.
7. In `src/utils/memory.irp.f`, comment out the `IRP_IF MACOS IRP_ELSE [...] do [...] enddo` blocks (there are two of them): because `/proc/self` files don't exist on MAC OS. In the same file, in the function  `get_total_available_memory` , replace `iunit = getUnitAndOpen('proc/meminfo', 'r')` with `iunit = 20`.
8. Return to folder `$QP_ROOT`
9. Dans etc/qp.rc, 
10. run `export QP_ROOT=/users/.../.../qp2` <- path of your quantum package folder
11. `source quantum_package.rc`  (if it does not work, re-open your terminal)
12. `./configure -c ./config/gfortran_macos.cfg` (I haven't tried the other configurations)
13. run `ninja`
14. While it runs (finger crossed), ask yourself why did you even tried to install quantum package on a mac os knowing that the first requirement is Linux OS... 
![th](https://github.com/user-attachments/assets/c99f9807-f771-4fa6-8abb-e113be475e8d)

In order to run calculations, do not use `./bin/qpsh`. use `source quantum_package.rc`.

**_Test:_**
In your calculations folder, create a file H2.xyz and write inside
```
2

H 0. 0. 0.
H 0. 0. 0.714
```

run the following commands:
```
source ~/qp2/quantum_package.rc
qp create_ezfio -b 6-31g H2.xyz
qp run scf
```

(Thank you chatGPT)

**Reply by scemama** [reply]
Hi, thanks for your message!
A simpler option is to install bash on macos ;-) 
Some time ago, I have tried to make QP work on MacOS, but I still had some issues. Is it working on your mac?

**Reply by t-diata** [reply]
If I knew... :')
At least I have been able to run a scf and a fci, I just installed it

**Reply by scemama** [reply]
OK, thanks, good to know!
